src/Voron/Impl/LowLevelTransaction.cs

```diff
@@ -260,215 +220,114 @@ private LowLevelTransaction(LowLevelTransaction previous, TransactionPersistentC
-        }
+            _env.WriteTransactionPool.Reset();
+            _dirtyPages = _env.WriteTransactionPool.DirtyPagesPool;
+            _transactionPages = new HashSet<PageFromScratchBuffer>(PageFromScratchBufferEqualityComparer.Instance);
+            _pagesToFreeOnCommit = new Stack<long>();
 
-        [Conditional("DEBUG")]
-        private void EnsureNoDuplicateTransactionId(long id)
-        {
-            EnsureNoDuplicateTransactionId_Forced(id);
+            InitializeRoots();
+            InitTransactionHeader();
         }
 
-        internal void EnsureNoDuplicateTransactionId_Forced(long id)
+        internal EnvironmentStateRecord CurrentStateRecord => _envRecord;
+
+        public bool TryGetClientState<T>(out T value)
         {
-            foreach (var journalFile in _journal.Files)
+            if (_envRecord.ClientState is T t)
             {
-                var lastSeenTxIdByJournal = journalFile.PageTranslationTable.GetLastSeenTransactionId();
-
-                if (id <= lastSeenTxIdByJournal)
-                    VoronUnrecoverableErrorException.Raise(this,
-                        $"PTT of journal {journalFile.Number} already contains records for a new write tx. " +
-                        $"Tx id = {id}, last seen by journal = {lastSeenTxIdByJournal}");
-
-                if (journalFile.PageTranslationTable.IsEmpty)
-                    continue;
-
-                var maxTxIdInJournal = journalFile.PageTranslationTable.MaxTransactionId();
-
-                if (id <= maxTxIdInJournal)
-                    VoronUnrecoverableErrorException.Raise(this,
-                        $"PTT of journal {journalFile.Number} already contains records for a new write tx. " +
-                        $"Tx id = {id}, max id in journal = {maxTxIdInJournal}");
+                value = t;
+                return true;
             }
+
+            value = default;
+            return false;
         }
 
         internal void UpdateRootsIfNeeded(Tree root)
         {
             //can only happen during initial transaction that creates Root and FreeSpaceRoot trees
-            if (State.Root != null)
+            if (_envRecord.Root != null)
                 return;
 
-            State.Initialize(root.State);
+            _envRecord = _envRecord with { Root = root.State };
 
             _root = root;
         }
 
+        internal void UpdateDataPagerState(Pager.State dataPagerState)
+        {
+            Debug.Assert(Flags is TransactionFlags.ReadWrite, "Flags is TransactionFlags.ReadWrite");
+            DataPagerState = dataPagerState;
+        }
+
+        internal void UpdateJournal(JournalFile file, long last4KWrite)
+        {
+            _envRecord = _envRecord with { Journal = (file, last4KWrite) };
+        }
+        
+        internal void UpdateClientState(object state)
+        {
+            Debug.Assert(_envRecord.ClientState == null || state == null || _envRecord.ClientState.GetType() == state.GetType(),
```

Comment:
I wonder if `Debug.Assert` is enough. I'm worried that we might have some scenario where we violate this assertion but we won't know about it because it was running in Release

Reply:
Not possible - that is the sort of thing that we check the _types_ that would always be the same between release & debug

Reply:
Yes, the type constraint is enough I think